### PR TITLE
Proposing a new algorithm for the Checker!

### PR DESCRIPTION
**Pull request by SylvainTakerkart**
OK, here is my code for the new checker... Note that I have added a mecanism for checking for the existence of the mandatory files. It can be run as the old one:

> cd ando
> python newchecker.py -v tests/dataset004/exp-Landing

What needs to be done:

* [ ] add new datasets to test it (take them from the ones available in https://github.com/INT-NIT/AnDO/pull/59)
* [ ] add proper error handling
* [ ] design unit tests (and probably add more datasets) to check that it does what it's supposed to!

### CODE REVIEW

File: ando/newchecker.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+def newchecker(input_directory):
+    """ Checks the ephys-BIDS validity of a data set.
+
+    The specifications that define what is checked by this function is available in the following document:
+
+
+    Parameters
+    ----------
+        directory : string
```
Comment (JuliaSprenger) [inline ando/newchecker.py:45]:
```suggestion
        input_directory : string
```

File: ando/newchecker.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    for ind, (root, dirs, files) in enumerate(os.walk(input_directory)):
+        # compute depth of the directory given as input
+        if ind == 0:
+            # remove ending / or \ if the input directory was given with it at the end
+            if root.endswith(os.path.sep):
+                root = root[:-1]
+            # count the number of / or \ in the directory name to estimate its "depth"
+            initial_depth = root.count(os.path.sep)
+        # estimate at which level we are
+        depth = root.count(os.path.sep) - initial_depth
```
Comment (JuliaSprenger) [inline ando/newchecker.py:70]: I think this code could be simplified and more robust by using the pathlib library. Then these lines could look like this 
```suggestion
            # remove ending / or \ if the input directory was given with it at the end
            root = pathlib.Path(root)
            # count the number of / or \ in the directory name to estimate its "depth"
            initial_depth = len(root.resolve().parents)
            
        # estimate at which level we are
        depth = len(pathlib.Path(root).resolve().parents) - initial_depth
```

File: ando/newchecker.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+            List of errors (empty if the data set is valid)
+    """
+
+
+    error_list = []
+
+    # walk through the directory tree using the os.walk function
+    for ind, (root, dirs, files) in enumerate(os.walk(input_directory)):
+        # compute depth of the directory given as input
+        if ind == 0:
```
Comment (JuliaSprenger) [inline ando/newchecker.py:63]: Is there a reason to do this within the loop and not outside using `input_directory`?

File: ando/newchecker.py (hunk 2 above)
Comment (JuliaSprenger) [inline ando/newchecker.py:70]: maybe depth could be even calculated using the pathlib `relative_to` method and counting the length of the result

**Review by JuliaSprenger — COMMENTED** [review]
The proposed algorithm is definitely much shorter and easier to understand. However we should discuss how flexible it is to be extended for future features like optional folders and dependencies between file(name)s.

**Comment by SylvainTakerkart** [comment]
@Slowblitz OK, it's all yours now! As discussed, make your list of checkboxes first ;)

**Comment by Slowblitz** [comment]
So as we discuss earlier I'll continue this PR, there is still some stuff to be done as describe below :

- [ ] Fix the TODO in the AnDOChecker.py
- [ ] add new datasets (One dataset per "rules" for each level)
- [ ] Add new unit test that check regexp and overall functioning  (this one should be done first if we want to do it in a "test driven development" way).
- [ ] Add the generate regexp function 
- [ ] add a proper Error handling 



Feel free to add task that I have maybe forgot =)

**Pull request merged**